src/webgpu/api/operation/rendering/stencil.spec.ts

```diff
@@ -200,33 +207,111 @@ g.test('stencil_compare_func')
         { stencilCompare: 'never', stencilRefValue: 2, _expectedColor: kBaseColor },
-        { stencilCompare: 'not-equal', stencilRefValue: 0, _expectedColor: kStencilColor },
+        { stencilCompare: 'not-equal', stencilRefValue: 0, _expectedColor: kGreenStencilColor },
         { stencilCompare: 'not-equal', stencilRefValue: 1, _expectedColor: kBaseColor },
-        { stencilCompare: 'not-equal', stencilRefValue: 2, _expectedColor: kStencilColor },
+        { stencilCompare: 'not-equal', stencilRefValue: 2, _expectedColor: kGreenStencilColor },
       ] as const)
   )
   .fn(async t => {
     const { stencilCompare, stencilRefValue, _expectedColor } = t.params;
     t.checkStencilCompareFunction(stencilCompare, stencilRefValue, _expectedColor);
   });
+
+g.test('stencil_fail_operation')
+  .desc(
+    `
+  Test that the stencil operation is executed on stencil fail. Triangle with stencil reference 2
+  fails the 'less' comparison function because the base stencil reference is 1.
+    - If the fail operation is 'keep', it keeps the base color.
+    - If the fail operation is 'replace', it replaces the base color with the last stencil color.
+
+  TODO: Need to test the other stencil operations?
+  `
+  )
+  .params(u =>
+    u //
+      .combineWithParams([
+        { operation: 'keep', _expectedColor: kBaseColor },
+        { operation: 'replace', _expectedColor: kGreenStencilColor },
+      ] as const)
+  )
+  .fn(async t => {
+    const { operation, _expectedColor } = t.params;
+
+    const depthSpencilFormat: GPUTextureFormat = 'depth24plus-stencil8';
+
+    const baseStencilState = {
+      compare: 'always',
+      failOp: 'keep',
+      passOp: 'replace',
+    } as const;
+
+    const failedStencilState = {
+      compare: 'less',
+      failOp: operation,
+      passOp: 'keep',
+    } as const;
+
+    const stencilState = {
+      compare: 'equal',
+      failOp: 'keep',
+      passOp: 'keep',
+    } as const;
+
+    const baseState = {
+      format: depthSpencilFormat,
+      depthWriteEnabled: false,
+      depthCompare: 'always',
+      stencilFront: baseStencilState,
+      stencilBack: baseStencilState,
+      stencilReadMask: 0xff,
+      stencilWriteMask: 0xff,
+    } as const;
+
+    const failState = {
+      format: depthSpencilFormat,
+      depthWriteEnabled: false,
+      depthCompare: 'always',
+      stencilFront: failedStencilState,
+      stencilBack: failedStencilState,
+      stencilReadMask: 0xff,
+      stencilWriteMask: 0xff,
+    } as const;
+
+    const passState = {
+      format: depthSpencilFormat,
+      depthWriteEnabled: false,
+      depthCompare: 'always',
+      stencilFront: stencilState,
+      stencilBack: stencilState,
+      stencilReadMask: 0xff,
+      stencilWriteMask: 0xff,
+    } as const;
+
+    const testStates = [
+      { state: failState, color: kRedStencilColor },
+      { state: passState, color: kGreenStencilColor },
```

Comment:
```suggestion
      // Always fails because the ref (2) is less than the initial stencil contents (1).
      // Therefore red is never drawn, and the stencil contents may be updated according to `operation`.
      { state: failState, color: kRedStencilColor },
      // Passes iff the ref (2) equals the current stencil contents (1 or 2).
      { state: passState, color: kGreenStencilColor },
```

Reply:
Thanks! I applied your suggestion.